### ISSUE TITLE
Operation not permitted in /Users/cm/Realm/realm-java/realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp line 81

### ISSUE THREAD

**Issue by GuaiShu** [realm/realm-java#3713]
Operation not permitted in /Users/zaki/fromgit/realm/realm-java/realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp line 92

I want create my realm file on external storage,when i use Realm.getDefaultInstance(),i get the exception.

In MyApplication.java:
        Realm.init(this);
        File directory = new File(Environment.getExternalStorageDirectory(), "RealmSamplesData");
        if (!directory.exists()) directory.mkdirs();
        RealmConfiguration config = new RealmConfiguration.Builder()
                .name("myrealm.realm")
                .directory(directory)
                .build();
        Realm.setDefaultConfiguration(config);

In MainActivity.java:
      mRealm = Realm.getDefaultInstance();

```

#### Version of Realm and tooling
Realm version(s): 2.1.1

Android Studio version: 2.2.2

Which Android version and device: Android 4.4, on PDA device
```

**Comment by GuaiShu** [comment]
@zaki50  Hi, This bug is not already fixed it?
![image](https://cloud.githubusercontent.com/assets/14322982/19793984/b8be717c-9d01-11e6-9e96-36a8d1d3046e.png)

**Comment by beeender** [comment]
@GuaiShu What kind of PDA device did you test it on?

**Comment by GuaiShu** [comment]
@beeender  We use the industrial PDA, the operating system is Android4.0 or Android4.4.
 The official website：http://www.idatachina.com/
![image](https://cloud.githubusercontent.com/assets/14322982/19796373/6cf5390a-9d16-11e6-8163-a86b977eb4a3.png)

**Comment by beeender** [comment]
@GuaiShu Are you in China? Could you please contact me through email mc@realm.io ? If we can access the device, things will be much easier.

**Comment by GuaiShu** [comment]
@beeender  Ok,I have sent you an email.
Have you tested on Android 4.4 and below? I also failed the Android Studio emulator (Android API: 19).

**Comment by beeender** [comment]
@GuaiShu Thanks a lot for the information. I can reproduce the issue on emulator (Android API: 19).

**Comment by GuaiShu** [comment]
@beeender ,@zaki50 , hello. When will the bug be repaired. I need a rough time.Thanks!

**Comment by beeender** [comment]
`mkfifo` returns `EPERM﻿⁠⁠⁠⁠` when creating fifo on external storage on Android 4.4 which is a different behavior than other recent Android versions. We will fix this soon.

**Comment by beeender** [comment]
Need https://github.com/realm/realm-core/pull/2309 and https://github.com/realm/realm-object-store/pull/197 to fix this.

**Comment by weitzj** [comment]
Seing the same error on a Sony Xperia Z1 Compact D5503, Android 5.1.1

**Comment by Zhuinden** [comment]
@weitzj that is a different issue, please check if all your Realm instances on background threads are closed with `finally { if(realm != null) { realm.close(); } }`

**Comment by PCabral90** [comment]
Hi all,

I am running the same issue and testing on Galaxy S7 version 6

File dbDirectory = getExternalFilesDir(null);

        Realm.init(this);
        RealmConfiguration config = new RealmConfiguration.Builder()
                .name("TestRealm.realm")
                .directory(dbDirectory)
                .build();
        Realm.setDefaultConfiguration(config);

        Realm realm = Realm.getDefaultInstance();

        realm.beginTransaction();
        realm.copyToRealmOrUpdate(...);
        realm.commitTransaction();
        realm.close();

12-22 16:26:08.551 29658-29658/? E/REALM_JNI: jni: ThrowingException 5, Operation not permitted in /Users/cm/Realm/realm-java/realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp line 92, .
12-22 16:26:08.551 29658-29658/? E/REALM_JNI: Exception has been throw: Unrecoverable error. Operation not permitted in /Users/cm/Realm/realm-java/realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp line 92
12-22 16:26:08.551 29658-29658/? D/AndroidRuntime: Shutting down VM
12-22 16:26:08.551 29658-29658/? E/AndroidRuntime: FATAL EXCEPTION: main
                                                   Process: compass.nta.ptims, PID: 29658
                                                   io.realm.exceptions.RealmError: Unrecoverable error. Operation not permitted in /Users/cm/Realm/realm-java/realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp line 92
                                                       at io.realm.internal.SharedRealm.nativeGetSharedRealm(Native Method)
                                                       at io.realm.internal.SharedRealm.getInstance(SharedRealm.java:205)
                                                       at io.realm.internal.SharedRealm.getInstance(SharedRealm.java:182)
                                                       at io.realm.RealmCache.createRealmOrGetFromCache(RealmCache.java:124)
                                                       at io.realm.Realm.getDefaultInstance(Realm.java:209)

And please don't ask about manisfest permissions because at this point i already triple checked!

**Comment by PCabral90** [comment]
I forgot to say 
Realm Version 2.2.1
Android Studio 2.2.3
Physical Device: Samsung Galaxy S7 Version 6.0.0

**Comment by Zhuinden** [comment]
@Pcabral90 but how about external storage write permissions?

**Comment by PCabral90** [comment]
Thanks for the quit answer!

@Zhuinden i have the permission

<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>

and i now that on version above 5 you need to check and grant but i manually granted.

**Comment by Zhuinden** [comment]
@PCabral90 actually it's above Android 6.0, I intended to ask specifically about *runtime* external storage permission. And if this happens in an activity or an application - specifically the question is primarily "what does `getExternalFilesDir(null).getAbsolutePath()` return.

**Comment by PCabral90** [comment]
@Zhuinden I have created a completely empty project and added only the Realm. The results are:

Physical Device: Android 4.4 works well
Emulator Device: Android 6.0.0 Galaxy S7 works well
Physical Device Android 6.0.0 Galaxy S7 **CRASH**

I tried the code in the Application class and Activity class the results are the same. I tried change the directory the different location and the result is the same. But i know this getExternalFilesDir(null) is correct because is working in emulator and Android 4.4. Even though the return result of the method for physical device Galaxy S7 is "/storage/emulated/0/Android/data/compass.realmtest/files"

**Comment by weitzj** [comment]
Just an overview when you need `WRITE_EXTERNAL_STORAGE` permissions:

- Writing to `getExternalFilesDir()` (API <= 18)
- Writing to an external directory **outside** your application(i.e. **not** `externalFilesDir`) (all API versions)

Some testcases:

- you are on API 23 and only write to `getExternalFilesDir` => no need to ask for permission (runtime or manifest)
- you are on API 17 and only write to `getExternalFilesDir` => need to have `WRITE_EXTERNAL_STORAGE` permission declared
- you are on API 23 and want to write to the external (secondary) storage outside your application => need to have `WRITE_EXTERNAL_STORAGE` permissions declared, need to ask during runtime.

**Comment by PCabral90** [comment]
@weitzj Thanks for the overview but the point is not about the permissions. I have the permissions i already check in runtime. And as you said on API 23 witch is Galaxy S7 you don't need the permission so the problem is not the permission and even if you need i already added and checked in runtime.

The problem continue to be the same the realm is not working on physical device Galaxy S7 but is working on the emulator.

**Comment by beeender** [comment]
Sorry for forgetting update status here. This issue should be fixed by #3904  . Although it has not been released yet, you can try with our snapshot release.

**Comment by PCabral90** [comment]
@beeender Thanks for the reply.

I have tried the snapshot and works.

Thanks.

**Comment by beeender** [comment]
Fix has been merged. will be released in `v2.2.2`.
